lib/sasl/sasl.js

```diff
@@ -60,41 +54,50 @@ Sasl.prototype.headerReceived = function(header) {
 };
 
 Sasl.prototype._processFrame = function(frame) {
+  var self = this;
   if (frame instanceof frames.SaslMechanismsFrame) {
-    var buf = new Builder();
     var mechanism = this.mechanism;
     var mechanisms = Array.isArray(frame.saslServerMechanisms) ?
       frame.saslServerMechanisms.map(function(m) { return m.value; }) : frame.saslServerMechanisms;
     if (!u.includes(mechanisms, mechanism)) {
       this.callback(new errors.AuthenticationError(
         'SASL ' + mechanism + ' not supported by remote.'));
     }
-    if (mechanism === Sasl.Mechanism.PLAIN) {
-      debug('Sending ' + this.credentials.user + ':' + this.credentials.pass);
-      buf.appendUInt8(0); // <null>
-      buf.appendString(this.credentials.user);
-      buf.appendUInt8(0); // <null>
-      buf.appendString(this.credentials.pass);
-    } else if (mechanism === Sasl.Mechanism.ANONYMOUS) {
-      if (this.credentials.user && this.credentials.pass) {
-        console.warn(
-            'Sasl ANONYMOUS requested, but credentials provided in endpoint URI');
-      }
-      buf.appendUInt8(0); // <null>
-    } else {
-      this.callback(new errors.NotImplementedError(
-          'Only SASL PLAIN and ANONYMOUS are supported.'));
+
+    if (mechanism === Sasl.Mechanism.ANONYMOUS && this.credentials.user && this.credentials.pass) {
+      console.warn(
+          'Sasl ANONYMOUS requested, but credentials provided in endpoint URI');
     }
-    var initFrame = new frames.SaslInitFrame({
-      mechanism: mechanism,
-      initialResponse: buf.get()
-    });
 
-    if (!!this._remoteHostname) initFrame.hostname = this._remoteHostname;
-    this.connection.sendFrame(initFrame);
+    this.handler.getInitFrameContent(self.credentials)
+      .then(function (initFrameContent) {
+        var initFrame = new frames.SaslInitFrame({
```

Comment:
that's exactly why. I'm not sure we want to start exporting our frames types. It's worth revisiting this decision now though since changing it later would likely be a breaking change.